v2/reader.go

```diff
@@ -54,10 +70,29 @@ func (r *Reader) readHeader() (err error) {
 
 // CarV1ReaderAt provides an io.ReaderAt containing the CAR v1 dump encapsulated in this CAR v2.
 func (r *Reader) CarV1ReaderAt() io.ReaderAt {
+	return r.carv1SectionReader()
+}
+
+func (r *Reader) carv1SectionReader() *io.SectionReader {
 	return io.NewSectionReader(r.r, int64(r.Header.CarV1Offset), int64(r.Header.CarV1Size))
 }
 
 // IndexReaderAt provides an io.ReaderAt containing the carbs.Index of this CAR v2.
 func (r *Reader) IndexReaderAt() io.ReaderAt {
 	return internalio.NewOffsetReader(r.r, int64(r.Header.IndexOffset))
 }
+
+// ReadPragma reads the pragma from r.
+// This function accepts both CAR v1 and v2 payloads.
+// The roots are returned only if the version of pragma equals 1, otherwise returns nil as roots.
```

Comment:
It makes me slightly uneasy that this API is *less* powerful with the newer and better CARv2.

Presumably we could always provide the roots, if for a version 2 we also consumed the inner carv1's header. I imagine that's impractical because there could be 1GiB of padding between them.

What worries me is that one could miss this line in the doc, write code that uses the roots list and happens to work with carv1 archives and is supposedly forwards compatible, but then would silently break in two weeks when it starts being fed carv2 archives.

One way to make this clearer would be to have a ReadVersion that just returns the version, and a separate ReadHeaders that can have a higher cost and always returns more information.

Reply:
> Presumably we could always provide the roots, if for a version 2 we also consumed the inner carv1's header. I imagine that's impractical because there could be 1GiB of padding between them.

That won't be reading the `Pragma`. the intention here is to explicitly read what the spec specifies as sections in a CART v2 file. According to the spec, if version is 2 then we must never have any roots.

> What worries me is that one could miss this line in the doc, write code that uses the roots list and happens to work with carv1 archives and is supposedly forwards compatible, but then would silently break in two weeks when it starts being fed carv2 archives.

Note if version is 2, roots are nil. That's an error in CAR v1 I think, since decoding a car v1 header with empty roots would return an empty slice. I'd love to understand the workflow you can think of that makes this function error-prone. There is something here to improve I am sure of it.

> One way to make this clearer would be to have a ReadVersion that just returns the version, and a separate ReadHeaders that can have a higher cost and always returns more information.

This means that function should only take `io.ReaderAt` or would swallow up `roots` if they are present in case version is 1.

The workflow I had in mind for this is that a user can probe what version they are dealing with and proceed to read the rest of the file using `carv2.Reader` _if_ the version is 2. Otherwise they would revert back to `carv1` libraries. I am hoping folks in ignite team would help us taylor this API to be most effective. For now what's here satisfies the requirement of being able to tell what version of a CAR file one is dealing with.

Reply:
I see. Then I think this could just be "give me the version", which will always be present? :)

If one has to look at the version and do something special for carv1, I think it's fine to then have to do something extra to get the roots list, such as use the carv1 decoder.

Reply:
That makes sense. thanks @mvdan so you think it's OK to swallow up roots if they are present in `ReadVersion`?

Reply:
Yeah, you'd just throw them away. You could even be smart about that: if you have a ReaderAt, when you encounter the roots list in dag-cbor, I imagine you could skip the entire roots list without fully decoding all of its elements. That can be an optimization TODO for now though.